### DISCUSSION TITLE
Chamber and exhaust fan macros

### DISCUSSION THREAD

**Discussion by Tiwatz** [xblax/flashforge_ad5m_klipper_mod#53] [Ideas]
I noticed that the chamber fan or the exhaust fan didn't start on my printer. Orca slicer puts out the g-code "M106 P2 S255" to start the aux/chamber fan. I did some research and the P value is not recognized by Klipper.

Also I am a little bit confused by the fan names but it is no big issue. The way it works on my 5M Pro original firmware is for external circulation, the lower (named internal_fan) fan pulls air out thru the lower filter and the upper fan (named external_fan) pulls air in.

For internal circulation, the lower fan is not spinning and the upper fan pulls air from inside the chamber from upper filter.
The selection of internal/external air in to printer is made by a servo.

In orca slicer we can set auxillary fan and exhaust fan. Not yet internal circulation but we can add "INTERNAL_AIR_CIRCULATION" in filament g-code. I don't know if it works on other slicers.

Added this to macro.cfg:
```
[gcode_macro INTERNAL_AIR_CIRCULATION]
description: Turn on internal air circulation
gcode:
  SET_FAN_SPEED FAN=external_fan SPEED=1
  SET_FAN_SPEED FAN=internal_fan SPEED=0
  SET_SERVO SERVO=my_servo ANGLE=90

[gcode_macro EXTERNAL_AIR_CIRCULATION]
description: Turn on external air circulation
gcode:
  SET_FAN_SPEED FAN=external_fan SPEED=1
  SET_FAN_SPEED FAN=internal_fan SPEED=1
  SET_SERVO SERVO=my_servo ANGLE=180

[gcode_macro STOP_AIR_CIRCULATION]
description: Turn off air circulation
gcode:
  SET_FAN_SPEED FAN=external_fan SPEED=0
  SET_FAN_SPEED FAN=internal_fan SPEED=0
  SET_SERVO SERVO=my_servo ANGLE=90

[gcode_macro M106]
rename_existing: M106.1
gcode:
  {% if params.P is defined and params.P|int > 0 %}
  {% if params.P|int == 1 %}
  RESPOND MSG="Fan does not exist!"
  {% endif %}
  {% if params.P|int == 2 %}
  SET_FAN_SPEED FAN=chamber_fan SPEED={(params.S|int / 255 )}
  {% endif %}
  {% if params.P|int == 3 %}
  SET_FAN_SPEED FAN=internal_fan SPEED={(params.S|int / 255 )}
  SET_FAN_SPEED FAN=external_fan SPEED={(params.S|int / 255 )}
    {% if params.S|int == 0 %}
    SET_SERVO SERVO=my_servo ANGLE=90
    {% else %}
    SET_SERVO SERVO=my_servo ANGLE=180
    {% endif %}
  {% endif %}
  {% else %}
  M106.1 S{params.S|int}
  {% endif %}
```

and then in the END_PRINT section we add STOP_AIR_CIRCULATION

I hope this helps anyone and would be glad for input since this is my first macro.

**Comment by consp** [comment]
> Also I am a little bit confused by the fan names but it is no big issue. The way it works on my 5M Pro original firmware is for external circulation, the lower (named internal_fan) fan pulls air out thru the lower filter and the upper fan (named external_fan) pulls air in.

Probably because neither @xblax nor me owns a pro 🥲 

> {(params.S|int / 255 )}

I don't know how jinja does it's integer rounding but if it is done in python and is strictly integer this code will always be 0 unless 255 is entered. Is the SPEED value supposed to be 0 to 1? Maybe make it float just to be sure?

**Reply by xblax** [reply]
The fan names for PRO fans are coming straight from the official klipper configuration. Afaik I didn't change them.

**Reply by consp** [reply]
That explains the naming scheme as well, probably fine in chinglish.

**Comment by Tiwatz** [comment]
I updated macro in the first post because in the old one the command ```M106 P3 S0``` would start the external fan.

**Reply by johncotten** [reply]
Just learned this, apparently GitHub isn't smart enough to recognize code when you paste it and doesn't leave the tab/spaces correctly...
"You can call out code or a command within a sentence with single backticks. The text within the backticks will not be formatted. You can also press the Command+E (Mac) or Ctrl+E (Windows/Linux) keyboard shortcut to insert the backticks for a code block within a line of Markdown. "
From: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

**Reply by consp** [reply]
<sub>Can also fixed by mod edit</sub>
